### ISSUE TITLE
Revisit train/validate/evaluate split

### ISSUE THREAD

**Issue by jemrobinson** [alan-turing-institute/ice-station-zebra#334]
### Problem / Goal

We've discussed a few times that we might want to change our default train/validate/evaluate split, particularly that we might like to move the holdout set and the evaluate set to after the end of any of our training data.

#### Current setup

```yaml
# This data is a holdout set not to be used until we are ready to produce final outputs (e.g. a paper)
# Data from 2013 and 2023
predict:
  - start: 2013-01-01
    end: 2013-12-31
  - start: 2023-01-01
    end: 2023-12-31

# This data is a test set used to compare fully-trained models (e.g. on the project leaderboard)
# Data from 2019
test:
  - start: 2019-01-01
    end: 2019-12-31

# This data is used for training the model
# Data from 1979-1999, 2001-2012, 2014-2018, 2020, 2022 and 2024
train:
  - start: null
    end: 1999-12-31
  - start: 2001-01-01
    end: 2012-12-31
  - start: 2014-01-01
    end: 2018-12-31
  - start: 2020-01-01
    end: 2020-12-31
  - start: 2022-01-01
    end: 2022-12-31
  - start: 2024-01-01
    end: 2024-12-31

# This data is used for validation during training (e.g. for early stopping etc.)
# Data from 2000 and 2021
validate:
  - start: 2000-01-01
    end: 2000-12-31
  - start: 2021-01-01
    end: 2021-12-31
```

#### Proposed setup

**Full dataset**

```yaml
# This data is a holdout set not to be used until we are ready to produce final outputs (e.g. a paper)
# Data from 2025 (1 year)
predict:
  - start: 2025-01-01
    end: 2025-12-31

# This data is an evaluation set used to compare fully-trained models (e.g. on the project leaderboard)
# Data from 2024 (1 year)
test:
  - start: 2024-01-01
    end: 2024-12-31

# This data is used for training the model
# Data from 1979-2021 (43 years)
train:
  - start: null
    end: 2021-12-31

# This data is used for validation during training (e.g. for early stopping etc.)
# Data from 2022 and 2023 (2 years)
validate:
  - start: 2022-01-01
    end: 2023-12-31
```

**Sample dataset**

```yaml
# This data is a test set used to compare fully-trained models (e.g. on the project leaderboard)
# Data from 2024 (2 months) [strict subset of full]
test:
  - start: 2024-01-01
    end: 2024-01-31
  - start: 2024-07-01
    end: 2024-07-31

# This data is used for training the model
# Data from 2020-2021 (24 months) [strict subset of full]
train:
  - start: 2020-01-01
    end: 2021-12-31

# This data is used for validation during training (e.g. for early stopping etc.)
# Data from 2022 and 2023 (2 months) [strict subset of full]
validate:
  - start: 2022-07-01
    end: 2022-07-31
  - start: 2023-01-01
    end: 2023-01-31
```

### What does good look like?

- [ ] Decision made on whether to change
- [ ] Change implemented in YAML files if needed

**Comment by louisavz** [comment]
Discussion during coworking - WIP - to be further discuss below. @Shaerdan @marianovitasari20 @jemrobinson

Have training data be linear in time (train before val before test data). 

Definition (refresher)
train: used for training models
validate: used for per epoch metric plotting
test: used to end-to-end test a trained model from checkpoint (e.g. for comparing models)
predict: just used when we write a paper (and as deliverable)

For example (for one config):
train:
  - start: null
    end: 2011-12-31
validate:
  - start: 2012-01-01
  - end: 2012-12-31
predict:
  - start: 2013-01-01
  - end: 2013-12-31
(thinking in terms of a paper showing results for extreme event)
Serdan: Extreme events would also be interesting to others, it's a very common question in conferences on how well our model perform.

Real-time model should look like this (as an example)
train:
  - start: null
  - end: 2020-12-31
validate:
  - start: 2021-01-01
  - end: 2021-12-31
test: (leaderboard - could get rid of this for the final model for one product, then shift train/validation data by a year)
 - start: 2022-01-01
 - end:  2022-12-31
predict:
  - start: 2023-01-01 
  - end: 2023-12-31 (this can be 2024) <- for shipping, this would be x-day forecasts out into the future (e.g., 2026.07.23 for a 7-day out forecast)

todo: Louisa add links to lit from other papers that uses linear windows for train/val/test/predict (also add MetNorway's and DMI's paper)
todo: Maria will do a little research and comment below

**Comment by marianovitasari20** [comment]
The original IceNet paper also handled this linearly (chronological split) since our project is based on it.

If we move to a linear split (train -> validate -> test -> holdout, in time order), I think that makes sense. The downside is that we lose the ability to pick specific extreme years.

I also thought about doing a rolling split (train up to year N, test on N+1, then move forward and repeat) to get more variety in test years. However, this would require retraining many times, which would be expensive.

**Comment by marianovitasari20** [comment]
I do agree that in many conferences people ask how good our model is at predicting extreme events. But I think it comes back to our priority, what kind of prediction do we want to prioritise for icenet-mp? Recent future prediction, or extreme event performance? Curious what others think, since this probably affects how we design the split.

**Comment by jemrobinson** [comment]
Summary of the above suggestions:

## Full dataset
- 43 full years

```yaml
# This data is a holdout set not to be used until we are ready to produce final outputs (e.g. a paper)
# Data from 2024
predict:
  - start: 2024-01-01
    end: 2024-12-31

# This data is a test set used to compare fully-trained models (e.g. on the project leaderboard)
# Data from 2023
test:
  - start: 2023-01-01
    end: 2023-12-31

# This data is used for training the model
# Data from 1979-2021
train:
  - start: null
    end: 2021-12-31

# This data is used for validation during training (e.g. for early stopping etc.)
# Data from 2022
validate:
  - start: 2022-01-01
    end: 2022-12-31
```

## Sample dataset (option 1)
- 4 full years

```yaml
# This data is a test set used to compare fully-trained models (e.g. on the project leaderboard)
# Data from 2023
test:
  - start: 2023-01-01
    end: 2023-12-31

# This data is used for training the model
# Data from 1979-2021
train:
  - start: 2020-01-01
    end: 2021-12-31

# This data is used for validation during training (e.g. for early stopping etc.)
# Data from 2022
validate:
  - start: 2022-01-01
    end: 2022-12-31
```

**Comment by jemrobinson** [comment]
> I do agree that in many conferences people ask how good our model is at predicting extreme events. But I think it comes back to our priority, what kind of prediction do we want to prioritise for icenet-mp? Recent future prediction, or extreme event performance? Curious what others think, since this probably affects how we design the split.

Great question. In my opinion, likely use-cases are for near-realtime prediction - e.g. predicting next week's sea ice from information available today. Do you agree @louisavz @scotthosking ?

**Comment by IFenton** [comment]
For the record, here are previous discussions on this topic, which contain some of the thinking on why we chose what we currently have
- #184 
- #101 
- #33

**Comment by IFenton** [comment]
A quick thought on the sample dataset. This change will make it almost twice the size of the one we currently use (2017-01 to 2019-01). I suspect that will make things slower to run, as well as being slower to download / taking up more space. We might be happy with that trade-off, but something to consider

**Comment by jemrobinson** [comment]
@IFenton: I think we could slim this down a bit by using some of the validation space for training (I think this is OK since validation/training are essentially an arbitrary split over the training data):

## Sample dataset (option 2)
- 2021-01-01 to 2023-01-31 (same size as current 2017-01-01 to 2019-01-31)

```yaml
# This data is a test set used to compare fully-trained models (e.g. on the project leaderboard)
# Data from 2023
test:
  - start: 2023-01-01
    end: 2023-01-31

# This data is used for training the model
# Data from 2021-2022
train:
  - start: 2021-01-01
    end: 2022-11-30

# This data is used for validation during training (e.g. for early stopping etc.)
# Data from 2022
validate:
  - start: 2022-12-01
    end: 2022-12-31
```

## Sample dataset (option 3)
- 2021-01-01 to 2023-06-30 (slightly larger than option 2 but ensures that each of test/train/validate have data from all seasons, if we don't mind interspersing train/validate data)

```yaml
# This data is a test set used to compare fully-trained models (e.g. on the project leaderboard)
# Data from 2023
test:
  - start: 2023-01-01
    end: 2023-06-30

# This data is used for training the model
# Data from 2021-2022
train:
  - start: 2021-01-01
    end: 2021-12-31
  - start: 2021-02-01
    end: 2022-06-30
  - start: 2022-08-01
    end: 2022-12-31

# This data is used for validation during training (e.g. for early stopping etc.)
# Data from 2022
validate:
  - start: 2022-01-01
    end: 2022-01-31
  - start: 2022-07-01
    end: 2022-07-30
```

**Comment by IFenton** [comment]
For the full dataset, do we want more than one year for validation / testing? If we're using the full 40+ years of data that's rather a small fraction. The examples in https://github.com/alan-turing-institute/icenet-mp/issues/184 use multiple years (and I think 10-20% often used as a rule of thumb)

**Comment by jemrobinson** [comment]
> For the full dataset, do we want more than one year for validation / testing? If we're using the full 40+ years of data that's rather a small fraction. The examples in [#184](https://github.com/alan-turing-institute/icenet-mp/issues/184) use multiple years (and I think 10-20% often used as a rule of thumb)

Given that we're mainly using 2000-202X at the moment, perhaps two years makes sense as approximately 10%? I think @louisavz was worried about the training data being too far away from the evaluate/hold-out data.

**Comment by louisavz** [comment]
I am/was (I can't decide!), but at some point we won't have the need for the "test" set, so it should be fine.

**Comment by louisavz** [comment]
> Great question. In my opinion, likely use-cases are for near-realtime prediction - e.g. predicting next week's sea ice from information available today. Do you agree [@louisavz](https://github.com/louisavz) [@scotthosking](https://github.com/scotthosking) ?

Yes, from what I'm seeing so far, in terms of using this pipeline for sea ice forecasting, our most likely use case would be predicting next week (possibly two) from information available today. I've been missing some discussions and haven't followed up yet, so @scotthosking please correct me if this is not the case.

**Comment by jemrobinson** [comment]
@AoifeHughes will do some studies on the importance of strict ordering and/or separation of training from prediction.

**Comment by louisavz** [comment]
I reached out to @ots22 and had a nice discussion today. We worked through this thought process that helps confirm our beliefs that we need to order the train / val / test / predict sequentially, with an additional realisation that a time gap between train and val datasets is ideal.

I made the case that the validation set should be ordered after the training data, not carved out of the middle of it, and that this holds on principle even if experiments show the forecast skill is numerically about the same either way.

The cleanest way to see it is the extreme case. Suppose we pick validation points at random across the time series, scattered throughout the training period. Because SIE is temporally smooth (yesterday's extent strongly predicts tomorrow's), every validation point has training points sitting right beside it on both sides. An expressive model doesn't have to forecast those validation points; it can essentially interpolate them from their neighbours, using data from after the target to reconstruct them. Validation loss looks great, but it's measuring the wrong thing because our deployment task is forecasting forward, where no future data exists.

Now group those points into bigger blocks such as weeks, months, and finally a whole year lifted out of the middle of the training set. Chunking makes the interpolation harder, but it doesn't remove the leak: the edges of that year are still flanked by training data on both sides, and an expressive enough model can still lean on the surrounding period (including the future side) to fill in the interior. A validation set with training data on both sides is structurally unable to represent the forward-only task we actually care about. This problem is made less obvious by the fact that we also have a validation set right after the training data.

Ordering the sets sequentially fixes this: validation strictly after training means the model must project forward into unseen time, which is the deployment task. This is why I'd argue for it even if an experiment shows no significant difference in forecast error. The validation set's job is to be an honest estimate of forward-forecasting skill. A val set embedded in training is biased in the optimistic direction, whether or not that bias happens to be large for this particular model and dataset. We shouldn't pick the split based on which one the current experiment can't distinguish. Instead, we should pick the one that's the right estimator, so we can trust it when the model or the data changes.

One refinement worth noting: even with sequential ordering, the start of the validation period is still influenced by the training data right before it, as December bleeds into January. A short gap between the two sets decouples that boundary correlation, giving an honest validation estimate, not inflated by leakage. This isn't meant to mimic deployment because operationally we'd forecast the next period with no gap at all.

**Comment by jemrobinson** [comment]
@louisavz: I think the proposal in the top-level comment now achieves this (except for the gap between train and validate). Are you happy with that or would you like to tweak it further?

**Comment by louisavz** [comment]
@jemrobinson I'm happy with that. The gap is nice to have but we don't have a lot of data, we'd get penalised either direction we adopt. Do we still want @AoifeHughes to run experiments?
